fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/data/RepaymentPeriod.java

```diff
@@ -0,0 +1,176 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanaccount.loanschedule.data;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import org.apache.fineract.infrastructure.core.service.MathUtil;
+import org.apache.fineract.organisation.monetary.domain.Money;
+
+@ToString
+@EqualsAndHashCode(exclude = { "previous", "next" })
+public class RepaymentPeriod {
+
+    @ToString.Exclude
+    private final RepaymentPeriod previous;
+    @Getter
+    private final LocalDate fromDate;
+    @Getter
+    private final LocalDate dueDate;
+    @Getter
+    private final List<InterestPeriod> interestPeriods;
+    @Setter
+    @ToString.Exclude
+    private RepaymentPeriod next;
+    @Setter
+    @Getter
+    private Money emi;
+    @Getter
+    private Money paidPrincipal;
+    @Getter
+    private Money paidInterest;
+
+    public RepaymentPeriod(RepaymentPeriod previous, LocalDate fromDate, LocalDate dueDate, Money emi) {
+        this.previous = previous;
+        if (previous != null) {
+            previous.setNext(this);
+        }
+        this.fromDate = fromDate;
+        this.dueDate = dueDate;
+        this.emi = emi;
+        this.interestPeriods = new ArrayList<>();
+        // There is always at least 1 interest period, by default with same from-due date as repayment period
+        getInterestPeriods().add(new InterestPeriod(this, getFromDate(), getDueDate(), BigDecimal.ZERO, getZero(), getZero(), getZero()));
+        this.paidInterest = getZero();
+        this.paidPrincipal = getZero();
+    }
+
+    public RepaymentPeriod(RepaymentPeriod previous, RepaymentPeriod repaymentPeriod) {
+        this.previous = previous;
+        if (previous != null) {
+            previous.setNext(this);
+        }
+        this.fromDate = repaymentPeriod.fromDate;
+        this.dueDate = repaymentPeriod.dueDate;
+        this.emi = repaymentPeriod.emi;
+        this.interestPeriods = new ArrayList<>();
+        this.paidPrincipal = repaymentPeriod.paidPrincipal;
+        this.paidInterest = repaymentPeriod.paidInterest;
+        // There is always at least 1 interest period, by default with same from-due date as repayment period
+        for (InterestPeriod interestPeriod : repaymentPeriod.interestPeriods) {
+            interestPeriods.add(new InterestPeriod(this, interestPeriod));
+        }
+    }
+
+    public void addInterestPeriod(InterestPeriod interestPeriod) {
+        interestPeriods.add(interestPeriod);
+        Collections.sort(interestPeriods);
```

Comment:
Why don't we use a sorted data structure instead? What about a TreeSet?

Reply:
No, sort is not needed anymore. It was part of old logic, I remove it.